Issue #1800: Make Scarf of shadows not affect the lenght of bounced spells. Also slight buff to player Kobolds, their spells bounce back one space further.

AlexanderPosch opened the pull request:
This pull request addresses Issue #1800:

Los restrictions (like the Scarf of shadows) reduce the length of lighting bolts/shocks reflected at walls.
![image](https://user-images.githubusercontent.com/49498206/114088621-7e46be00-98b5-11eb-8e36-6ef4d632810e.png)

The code checks how long a bounced spell would have been if los had not been restricted and then increases the range of bounced the spell.  (Kobolds nightstalker mutation does not count as a los restriction)

This change does not allow player to affect tiles outside of the los-radius (unless you wear a scarf of shadows and the "Robe of Night" together then it is possible).

Additionally I also gave a slight buff to player kobolds, their spells bounce back one further. Since for kobolds all reflected spells(Shock,Lighting_Bolt,Fire_Ball,Starburst) have range 4 and this results in this:
![image](https://user-images.githubusercontent.com/49498206/114081070-69b1f800-98ac-11eb-9389-7a73c30de400.png)
rather than this:
![image](https://user-images.githubusercontent.com/49498206/114086296-bd274480-98b2-11eb-9a11-574097df28d7.png)
Also this change does not allow players to affect tiles outside of the los-radius (unless you wear a scarf of shadows and the "Robe of Night" together then it is possible).

I think it looks nicer and more symmetrical this way and I don't think it is very unbalancing.

ebering commented on crawl-ref/source/beam.cc:
```diff
@@ -837,7 +837,19 @@ void bolt::bounce()
+    	{
+    	extra_range_used -= spell_range(boltcopy.origin_spell, boltcopy.ench_power,true,true)-range;
+    	
+    	if  (you.species==SP_KOBOLD)
```
this should check for the nightstalker mutation instead of the species directly.

ebering commented on crawl-ref/source/beam.cc:
```diff
@@ -837,7 +837,19 @@ void bolt::bounce()
     extra_range_used += 2;
-
+ 
+    //Keep lenght of bounced Spells constant despite reduced los (scarf of shadows,Robe of Night)
```
typo

ebering commented on crawl-ref/source/spl-util.cc:
```diff
@@ -985,10 +985,19 @@ int spell_power_cap(spell_type spell)
+    int los_cap  = (int)you.current_vision;
+    if (ignore_los)
+    	{ 
+    	if (you.species!=SP_KOBOLD)
+    		los_cap  = (int)you.normal_vision;
+    	else
+    		los_cap  = (int)you.normal_vision-3;
```
This should check the nightstalker mutation directly and use its calculation.

ebering reviewed (changes requested):
Hello and welcome to contributing to crawl!

Thank you for this. In addition to the in-line changes, this needs to prevent out-of-los attacks when using shadows and Night together. Rare though it may be, we have spent quite a lot of effort moving away from out of los attacks in most cases.

AlexanderPosch commented:
Thanks for your quick reply and comments.  I'll try to fix the remaining issues.

AlexanderPosch commented:
> ...this needs to prevent out-of-los attacks when using shadows and Night together. Rare though it may be, we have spent quite a lot of effort moving away from out of los attacks in most cases.

Shock and Lighting Bolt can already attack creatures outside of los by bouncing around a corner:
![image](https://user-images.githubusercontent.com/49498206/114106307-c4f4e200-98ce-11eb-8461-641d24769608.png)
Wearing the Robe and scarf would allow attacking like this:
![image](https://user-images.githubusercontent.com/49498206/114106975-46993f80-98d0-11eb-9b18-957efc4b8820.png)


Other spells that can attack creatures outside los are:
- Kills outside los_radius: Gell's Gravitas, Fulminant Prism (in this case it makes sense), Fire Storm (needs luck or los-1), Fire Ball (needs los-2)
- Kills enemy around a corner: Spark, Lightning Bolt, LRD, Fire Storm, Fire Ball

Should I aim to only remove the new possibility for out-of-los attacks or all out-of-los attacks using bouncing? 
I think it might be easier if removing out-of-los spell-attacks were done independently of the bouncing mechanic for all spells by checking if affected targets are within los (or los_radius) of the player.

AlexanderPosch commented:
The newest change to the code prevents the out-of-los attacks that wearing the "Robe of Night" and a "Scarf of Shadows" allowed. Bolts going around corners can still hit enemies but this was already possible before this pull request.  
If you want _(and I figure out how multiple vectors with scalars)_ I should not be to hard to change the code to also prevent these "around corners" shots.

AlexanderPosch commented:
What should the range in the spell description show? The "normal range" or the "range restricted to los" or a mixture (like this @--->~~~...)?

AlexanderPosch commented:
### Summary of all changes:

- Exchanged kobold species checks for nightstalker mutation checks. Fixed typos
- Prevented bounced bolts from going outside of a radius of you.current_vision. This prevents the the out-of-los attacks that wearing the "Robe of Night" and a "Scarf of Shadows" allowed.
- Changed the spell_range_string, range_string so that the actual range of a spell can be seen the in spell description

ebering commented:
> I think it might be easier if removing out-of-los spell-attacks were done independently of the bouncing mechanic for all spells by checking if affected targets are within los (or los_radius) of the player.

This would be the much better solution. It's far enough out of scope for this PR that I will hunt it down separately when I do the merge. For now you can remove the special case you added (Sorry about sending you in a circle on this!).

I've had a rough weekend, and less crawl time than I had hoped so this isn't a complete review, just dropping a note to let you know where things are. I hope to get to one soon.

@semitonal mentioned that the Kobold special case for bounce range wasn't something she was interested in and I agree. While the geometric symmetry is pleasing, the preferred crawl style is mechanical symmetry.

AlexanderPosch commented:
Thanks for the update :). It makes sense that mechanical symmetry should come first.
An idea for keeping mechanical symmetry while also implementing some geometric symmetry might be to just completely ignore the night-stalker mutation in the bouncing mechanic. But I can see why this might also be undesirable.

ebering commented:
Thanks for the patch! I ended up tweaking things quite a bit, you can see the final result in the commit referencing this PR.

How would you like to appear in CREDITS.txt?

AlexanderPosch commented:
Thank you, It does seem a lot cleaner after your tweaking :) .
What ever is typical (Alexander Posch is fine). But since my contribution was rather little I would also be ok with not appearing.
